java/dictionary-renderer-graphics-overlay/src/main/java/com/esri/arcgisruntime/sample/dictionaryrenderergraphicsoverlay/MainActivity.java

```diff
@@ -0,0 +1,235 @@
+import com.esri.arcgisruntime.mapping.view.GraphicsOverlay;
+import com.esri.arcgisruntime.mapping.view.MapView;
+import com.esri.arcgisruntime.symbology.DictionaryRenderer;
+import com.esri.arcgisruntime.symbology.DictionarySymbolStyle;
+import org.xmlpull.v1.XmlPullParser;
+import org.xmlpull.v1.XmlPullParserException;
+import org.xmlpull.v1.XmlPullParserFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class MainActivity extends AppCompatActivity {
+
+  // permission to read external storage
+  private final String[] reqPermission = new String[] { Manifest.permission.READ_EXTERNAL_STORAGE };
+
+  private MapView mMapView;
+
+  private GraphicsOverlay graphicsOverlay;
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_main);
+
+    // get the reference to the map view
+    mMapView = findViewById(R.id.mapView);
+
+    ArcGISMap map = new ArcGISMap(Basemap.createTopographic());
+    mMapView.setMap(map);
+
+    // for API level 23+ request permission at runtime
+    if (ContextCompat.checkSelfPermission(this, reqPermission[0]) == PackageManager.PERMISSION_GRANTED) {
+      applyDictionaRendererToGraphics();
+    } else {
+      // request permission
+      int requestCode = 2;
+      ActivityCompat.requestPermissions(this, reqPermission, requestCode);
+    }
+  }
+
+  @TargetApi(24)
+  private void applyDictionaRendererToGraphics() {
+    graphicsOverlay = new GraphicsOverlay();
+    // graphics no longer show after zooming passed this scale
+    graphicsOverlay.setMinScale(1000000);
+    mMapView.getGraphicsOverlays().add(graphicsOverlay);
+
+    // create symbol dictionary from specification
+    DictionarySymbolStyle symbolDictionary = new DictionarySymbolStyle("mil2525d",
+        Environment.getExternalStorageDirectory() + getString(R.string.mil2525d_stylx));
+
+    // tells graphics overlay how to render graphics with symbol dictionary attributes set
+    DictionaryRenderer renderer = new DictionaryRenderer(symbolDictionary);
+    graphicsOverlay.setRenderer(renderer);
+
+    // parse graphic attributes from a XML file
+    List<Map<String, Object>> messages = parseMessages();
+
+    // create graphics with attributes and add to graphics overlay
+    messages.stream()
+        .map(MainActivity::createGraphic)
+        .collect(Collectors.toCollection(() -> graphicsOverlay.getGraphics()));
+
+    // once view has loaded
+    mMapView.addSpatialReferenceChangedListener(e -> {
+      // set initial viewpoint
+      mMapView.setViewpointGeometryAsync(graphicsOverlay.getExtent());
+    });
+  }
+
+  /**
+   * Parses a XML file and creates a message for each block of attributes found.
+   */
+  private List<Map<String, Object>> parseMessages() {
```

Comment:
This method could benefit from some in line comments to say what each part is doing

Reply:
done. Added comments